chromedriver as package

scriptworld-git opened the pull request:
removed installing step for downloading chromedriver
updated requirements
changed chromedriverpath in settings.py if not already exist in assets folder

timgrossmann commented:
@scriptworld-git Could you give a quick heads up how this works?
How does updating the chrome driver work?
Do they always have the latest version?
By whom is the dependency maintained?

scriptworld-git commented:
@timgrossmann  
The package is really simple, you can check the source code on my repositories.  
It just checks wether you are on a Linux, Mac or Windows Platform, and provides the path of the correct executable.

Since all 3 executables are delivered when installing the package, the path will be an absolutepath to the package folder (which contains the files).

I also wrote a script which is scraping the chromedriver webiste, checks for new versions, downloads and pushes them into the package (git repo AND pypi package). (I'll post this one on my gists soon)  
It's running daily on a server I rent.

As long as I am breathing, this package will be up to date, since its literally no effort for me to maintain it.

To update your driver simply go with `pip install -U chromedriver-py`.

If you got any doubts left, let me know so we can clear them out. If you any more features for this package, I'll implement them !

Well we can't get rid of the chromedriver... but at least its easier to integrate for now ! For example if you wanna make instapy a python package.  
It should also be possible to write better Unit Tests since Travis is able to run the driver aswell (but I am not quite sure yet, I will check that !

timgrossmann commented:
@scriptworld-git You're the man, sir!
Really appreciate that, didn't know that it could be this "easy" to take some of the work it takes to setup InstaPy and just put it into a package.

This is awesome!

Yeah I think I could sleep better if you could upload/post that gist so people know what's happening.
From the current view, you can't insert any weird or malignant code that way since it's only the path to the chrome driver (unless you write your own chromdriver file which implements the methods used in the tool... 😅)

Sorry, by saying getting rid of the chrome driver I actually meant the step people have to take to download it and put it into the assets folder by hand.

> For example if you wanna make instapy a python package

Absolutely true, this was the one thing I had no idea on how to make it installable via PyPi.
Definitely learned a lot of new stuff today 😄 

> It should also be possible to write better Unit Tests since Travis is able to run the driver aswell (but I am not quite sure yet, I will check that !

Super interesting!

> f you any more features for this package, I'll implement them ! 

Haha, always happy to have more people who are passionate about our tool and want to make it a bit better! 🙏 

@sionking @CharlesCCC @uluQulu What's your guys opinion about this approach?

uluQulu commented:
Wow That's impressive, @scriptworld-git but you have touched a red line 😉

I had the same questions in mind as in @timgrossmann's https://github.com/timgrossmann/InstaPy/pull/2947#issuecomment-422424964. 

Well. We already use lots of packages from PyPI BUT **webdriver** is not like them it is a basic bone for this program AND has the highest priority after _selenium_ package.

Chromedriver is a big EXECUTABLE package and IT is not possible to see WHAT IT HAS by reading inside that BINARY.
And I would be careful WHILE putting a **webdriver** into my assets folder, it must be the official one and/or FULLY secure alteration by trustworthy verification.

I don't mean @scriptworld-git's project is not trustworthy.
I hope you get my point on WHY I said you touched the red line 😉 @scriptworld-git 

I will offer 2 options to @scriptworld-git:
1-) Can you add some MD5 or SHA hash kind of verification and explicity VERIFY it after each pip downloads SO THAT people can use it COMFORTABLY?

2-) Or better, I always thought of doing it for this problem- auto getting the webdriver and leave the pain behind for the user.
>I also wrote a script which is scraping the chromedriver webiste, checks for new versions, downloads and pushes them into the package (git repo AND pypi package). (I'll post this one on my gists soon)
It's running daily on a server I rent.

Can you share the code you use to get **webdriver**, @scriptworld-git?
So that it can be directly implemented into **InstaPy** and have people feel 100% secure 🙂

My point is offering auto-get webdriver service for both **chromedriver** and **geckodriver**
And having parameters like,
+ `"latest"`: version that is the last stable release of the webdriver
+ `"compatible"`: version that is the last stable version that is compatible with the current Web browser version

Having considered that these days web browser auto-update themselves [most of times], then the parameters above will be suffient.

----

Cheers 😁
----

Nocturnal-2 commented:
Thanks to everyone who are working hard to make this a better project! 

This could make things easy for many users. But I am wondering whether this will work for Raspberry Pi users. I learn from one of the discussions [here](https://github.com/timgrossmann/InstaPy/issues/2838#issuecomment-418333161) that recent versions of chromedriver are not supported in Raspberry Pi 3 models which many people use. 

I am not expert, I apologize if this doesn't make any sense.

scriptworld-git commented:
@timgrossmann thanks ! well it took me some hours to figure out how to publish a package on pypi without messing anything up :D  
It's more difficult than I thought in the beginning!
 
[chromedriver-updater.py](https://gist.github.com/scriptworld-git/e55b2b618ea2e6b1c2006b473e6ad755)

The script is not 100% finished yet, still need to download the binaries, replace and run 2 bash commands. But it's already checking wether there is a new version compared to current version, and gets all urls to download the new ones.

@uluQulu Thanks for mentioning the security part. I didn't really think about it to be honest ! But you are right, I could modify the binaries.

I like your idea of specifying a version (or saying "latest") and if you start instapy it checks wether there are binaries in assets folder, if not, downloading only the one for current operating system with the given version (which is really easy), or if its set to latest, it will check wether there is a newer version and replaces it with the current one.

That way @Nocturnal-2 doubt would also be solved, because he could specify version x.y in the script and would download the current one which runs on raspberry pi.

But, since I still wanna maintain the python package, @uluQulu I didn't really understand your solution with verification. Could you explain it a little bit more so I can still implement it for the package maybe?

timgrossmann commented:
@scriptworld-git Do you think you could somehow integrate this with InstaPy directly like @uluQulu mentioned?

scriptworld-git commented:
@timgrossmann do you want it as i described in my post? that its automaticly downloaded for the right platform and right version?

I thought about it today, this is possible, but i dont know if you are able to convert instapy into a package then.  
Im not sure wether a package can download files. I also dont know where to store the file, since the assets folder is outside of the actual instapy directory. This would work with the current setup, but not as a package (or maybe docker, im not sure).

timgrossmann commented:
@scriptworld-git That would be the perfect case!

I think my idea is, if I understood @uluQulu right, the quite similar to his.
Would be nice if the flow could look like this:

On start of InstaPy, check for chrome driver.
If it is not there, get the ("latest" or "compatible", even I don't know how to find out the "compatible" one) version for the current system.
If there is a chrome driver, check if it's one of that 'specific' ones.
If it is, use that one.
If it's a downloaded one, check if it's still the latest version (or the compatible) one.

I think the idea that it's "only" downloaded or updated once the session is triggered resolves the problem package building. Or am I missing something?

Hopefully this was somewhat clear, sorry if I miscommunicated something.

> do you want it as i described in my post?

If I understood your post right, then yes, that would be great!

Thank you very much

scriptworld-git commented:
@timgrossmann @uluQulu I got a problem starting instapy.... get some database errors...

is it possible that you might quickly test my code ?  
it should work, I tested the downloading and unzipping part without instapy, but I couldn't test the whole project.

Workflow is just as discussed :)

Waiting for your Feedback! I can still change the behaviour :)

uluQulu commented:
Hi there!
I just had a quick look at your code, @scriptworld-git and looks clear 🐬 and I will try to read its details as gets some free time.
I will share a few ideas 
@timgrossmann, to get the "_compatible_" version one can read the release notes of the webdriver where it says, e.g. compatible with Chrome `67`-`70` and getting the max compatible version of the webdriver.
IT will be great parameter cos in some cases users does have a web browser older than the latest version and always getting the latest version of the webdriver might make a problem.
And a great idea came from @scriptworld-git is having an option to get the specific version of the webdriver WHICH will be so handy.
so the parameters can be,
+ "_latest_"
+ "_compatible_"
+ "{_custom_}"

>Could you explain it a little bit more so I can still implement it for the package maybe?

Uhm, maybe you can include a quick TRUSTWORTHY [3rd-party] tool to verify the integrity of the files in your project [package] and include it in documentation and make it easily accessible.
If you solve that matter, then your idea is gonna be amazing.
You can have extra smart parameters in calls [the ones we will use in here], support other webdrivers [geckodriver, etc.], etc.

Also, @scriptworld-git can you rename **chromedriver_util.py** to sth generic, e.g. **webdriver_util.py** so in the future support for e.g. _geckodriver_ can be added?

----

Cheers 😁
----

scriptworld-git commented:
@uluQulu Thanks :)

Getting the latest and specific version is possible right now. A user just have to pass the version to the constructor of instapy (im gonna write it to the readme when its all done).  
A user can also disable auto updating the browser and just download them on their own.

Im gonna rename the file and split it out a little bit nicer since its all in one function right now (because I used my gist ... sorry about that :D)

About the compatible version.... this is a little bit harder to scrape...  
I would suggest having a Gist where we paste the current compatible version for some cases into a simple file and every time we search for updates, we first of all download the gist, parse it, and see which is the latest compatible version (if version is set to compatible)

Even user who don't always download the newest project files, would still get a driver update if its compatible for their system.

timgrossmann commented:
@scriptworld-git If this is going to work as expected, it will be absolutely incredible and big step towards finally providing a simple way to install InstaPy for non technical people.

Next up probably is:
- checking if the dependencies are installed, if not install them automatically
- Building a new GUI with the api.json file as a base so we don't have to modify it every time something new is added

In the best case, people soon are able to use InstaPy, at least, without the CLI.
They can simply double click the quickstart.py file in order to start it. They "only" need to do the modification and customisation in the quickstart file without ever touching the CLI for installing dependencies or running the quickstart file.

scriptworld-git commented:
@timgrossmann I just split everything up into some functions, so its better readable... nothing fancy but still nicer :D

My question would be, could you guys test the code to approve it? Unfortunately I wasn't able to test it in the whole project ...

It should work as expected now, so feel free to merge it if your tests worked fine! 

About removing the CLI:  
Did you ever think about a bash / bat script? Thats really handy, you could just run pip install and start quickstart.py with python, and the user only has to doubleclick bash or bat (windows) file.

BUT.... i think thats not good, since the user still has to know a little bit about python to modify quickstart.py. If they wont read the docs, they will spam the issues, just because their 'one time double click bot' doesnt work... You know, people are lazy :)  
And in the current setup, you encourage people to learn new stuff... I guess thats the best feature of this open source bot !

BTW: I will still keep chromedriver package up to date, maybe ill find a way to approve it in the future.

timgrossmann commented:
@scriptworld-git Thank you very much for all that work you put into this!
I still think the python package is a good idea, so I'm glad you're keeping it.

I hope I can find some time to test your PR tomorrow.

Sorry for the delay.
Best
Tim

scriptworld-git commented:
@timgrossmann @uluQulu I just had a bug, where my bot (also written with chromedriver on python) just opened a blank white page and would proceed.

This was because of chromedriver version 2.42... when switching back to 2.38, my bot started working again!

I don't know if this issue is already known here, but since its related to this topic I thought it would be nice to mention it :)

timgrossmann commented:
@scriptworld-git Awesome and quick as heck 👍 

I'll test it this evening.
Thank you!

scriptworld-git commented:
@timgrossmann 

`chromedriver_min_version = 2.36`

You might want to delete this line in settings.py. I don't know if it's still needed for other validation.

It's been replaced with `chromedriver_specific_version`.

No problem! Glad to help out :)

timgrossmann commented:
@scriptworld-git Sorry, I'll move that checking to tomorrow morning, I'll be at university then, having access to all three, Linux, Mac and Windows for testing.

timgrossmann commented:
@scriptworld-git I check it on Ubuntu, MacOS and Windows but I always get the same result.
It's always displaying the "ensure chromedriver is installed at path" error message.
It seems like it's not trying to download the latest version or anything we definitely need some logging for this.

Something like:
- "Checking for chromedriver..."
- "Latest chromedriver already present..."
- "Retrieving latest chromedriver..."

etc.

Do you think you could add some more progress logging just so we are able to see what's going on a, a bit better?

Thank you very much, it feels like we are close to actually being able to use this!

scriptworld-git commented:
@timgrossmann Just added some logging Information :) Just show me a logger run ! 

I'll try to get it going on my machine aswell. This Database bug can't be that hard to fix :)

scriptworld-git commented:
@timgrossmann So i fixed all bugs now.

We got only one problem left: chromedriver tells me, I got no permission to execute the downloaded chromedriver version... maybe I just messed around a little too much, please try it on your own, and on different environments.

Otherwise we have to check wether we can adjust the permissions BUT I think this will be hard with a sudo user since otherwise a python script could download a virus and just run it... without root permissions :)

Well if this doesn't work we have to go back to chromedriver package since this works totally fine on my Bot :)  
Maybe someone could help me introduce some verification into the package if this won't work.

The script is downloading the latest version. If a desired version if given, he will check if the desired version is installed, if the version is higher or lower, it will install the desired one.

timgrossmann commented:
@scriptworld-git Awesome work! 
Thank you for the fixes. I'll try it out on all three platforms tomorrow just to make sure.

If we encounter more problems like the one you mentioned, I think you're right and we should think about building on your project to make it work but still be "on the safe side" with some verification on your end.

Thank you

uluQulu commented:
Thanks @scriptworld-git for this great dedicated work!
I have just had a quick glance into your code and it is much clearer now.
And then downloaded your fork and run a test in my Windows 10 Pro amd64 machine.
Here are the notes I have for you:
+ Having no any webdriver [chromedriver in this case] in my assets folder I was expecting the program to auto-get it and it failed with lots of `InstaPyError("ensure chromedriver is installed at {}")`.
Why?
Cos it failed to download a webdriver executable.
Why failed?
Cos I did not install the `bs4` package  [which was in the requirements.txt correctly - my mistake]

But the problem there is, it does misinform.
You should work on this block much better in the instapy.py file, esp. its `except` block:
```python
        if update_driver:
            try:
                driver_update(driver_version, self.logger)
            except:
                self.logger.error("Error updating webdriver")
```
E.g. simply you can do this:
If there is a viable webdriver executable exists [which InstaPy can use] in the assets folder, then pass with the "error with updating webdriver" message
But if there is not any usable webdriver executable in the assets folder, then print "You have not any chromedriver to run!" and provide a `raise` for the user to solve it, at least.
Those are just quick thoughts, it can be handled much smarter.

Fine. after I installed `bs4` from pip and it worked flawlessly, and you have done a good job!
One thing I can also say if you could do,
+ store the webdriver files at this name-style: `chromedriver_2-41.exe` rather than `chromedriver.exe`
+ and keep the latest 3 or 4 chromedrivers [for user to quickly switch]
I think since specific version is possible, then it can always download but instead of downloading, it can load an already existing file in no time.
E.g. think about a tester quickly switching between the versions, etc.

I have no Linux to test it  but I think sudo should not make a problem. 
Cos if it makes a problem, it will make it in many places, not just that part.
Why? - some users they already run with that command for whom sudo makes a problem [privilege issues].

----

Cheers 😁
----

uluQulu commented:
My notes out of code review:

[-] `driver_update()` and `update_webdriver()` function names are so similar 😟 which may cause a bit confusion for the future maintainer(s)


-----

Cheers 😀
----

uluQulu commented on instapy/instapy.py:
```diff
@@ -216,11 +216,17 @@ def __init__(self,
         Settings.chromedriver_desired_version = str(driver_version)
-        if update_driver:
+
+        if update_driver and not use_firefox:
```
if `update_driver=False` and there is no any webdriver available to use then, you can say this:
"Hey you have turned off webdriver auto update and you have no any webdriver executable to use.
Either turn on auto webdriver update or manually put some webdriver into the assets folder 🤓"

But with a shorter sentence  😂😄😁

scriptworld-git replied:
Also thought about that BUT if you turn off auto_update, you know what you have done since it's on purpose. If you read there is no web driver, you should know that you could turn it back one :D 
I was commenting on your comment :D wait I'll change this message !

scriptworld-git commented:
@uluQulu Thanks for this Feedback! I'm sorry if this work seems a little Spaghetti to you... That happens when building a one File script into a big Project ... :D

> You should work on this block much better in the instapy.py file, esp. its except block:  

I adjusted the except block. It will now print out better logs and tries to access a local chromedriver version.

> Also, "Current Webdriver Version: 0.0" message is a bit noisy.  

Changed this Error Message, it's much clearer now :)

> I have no Linux to test it but I think sudo should not make a problem.  

Let's wait for @timgrossmann for this review since it doesn't work on Mac... at least for me, could be that I messed up the permissions somehow!  
Glad to hear that this worked on windows :)

I changed the function names and cleaned up some constants. Also specified that you have to pass the number as a string. (also in README) 
There wouldn't be any other chance to make a difference between 2.4 and 2.40 (still think that Chromedriver made some huge versioning mistake here :D Shouldn't it be 2.04 ? Anyways, I tested and it worked if you pass a string.)

> imports are spread around the code- why don't you add them to the top of file? same for some constants inside functions....  

I learned that this is much better for performance, since python won't load this package into RAM if it doesn't reach the if statement (or function). But yeah, just a different style, I guess this won't matter overall !

>  keep the latest 3 or 4 chromedrivers [for user to quickly switch]  

Also thought about that but I didn't want to implement, since after a while this folder is polluted with old versions. Downloading the driver takes around 2-5 seconds for me. Thats not a big deal :D  
If you still want this Feature, you have to be patient tho, since I am really busy with work and my own project sadly... I'm not even using InstaPy this is why I prefer programming for another project :) But I really wanna get this finished in a way that this feature is useable!

timgrossmann commented:
@scriptworld-git I'm absolutely excited about this PR soon getting merged and this feature being available to all the people new to InstaPy!

Awesome that it works on Windows.

MacOS won't even download the chromedriver

Linux throws the following error when executing it:
```
INFO [2018-09-26 13:36:35] [contacting.john.doe]  Current webdriver version: 2.42
INFO [2018-09-26 13:36:36] [contacting.john.doe]  Latest version available is: 2.42
INFO [2018-09-26 13:36:36] [contacting.john.doe]  Desired webdriver version: 2.42
ERROR [2018-09-26 13:36:36] [contacting.john.doe]  Message: 'chromedriver' executable may have wrong permissions. Please see https://sites.google.com/a/chromium.org/chromedriver/home
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/common/service.py", line 76, in start
    stdin=PIPE)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
PermissionError: [Errno 13] Permission denied

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/InstaPy_3/instapy/instapy.py", line 351, in set_selenium_local_session
    chrome_options=chrome_options)
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/chrome/webdriver.py", line 73, in __init__
    self.service.start()
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/common/service.py", line 88, in start
    os.path.basename(self.path), self.start_error_message)
selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable may have wrong permissions. Please see https://sites.google.com/a/chromium.org/chromedriver/home

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/common/service.py", line 76, in start
    stdin=PIPE)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
PermissionError: [Errno 13] Permission denied

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/InstaPy_3/instapy/instapy.py", line 351, in set_selenium_local_session
    chrome_options=chrome_options)
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/chrome/webdriver.py", line 73, in __init__
    self.service.start()
  File "/usr/local/lib/python3.5/dist-packages/selenium/webdriver/common/service.py", line 88, in start
    os.path.basename(self.path), self.start_error_message)
selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable may have wrong permissions. Please see https://sites.google.com/a/chromium.org/chromedriver/home


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "quickstart.py", line 22, in <module>
    multi_logs=True)
  File "/root/InstaPy_3/instapy/instapy.py", line 236, in __init__
    self.set_selenium_local_session()
  File "/root/InstaPy_3/instapy/instapy.py", line 355, in set_selenium_local_session
    Settings.chromedriver_location))
instapy.instapy.InstaPyError: ensure chromedriver is installed at /root/InstaPy_3/assets/chromedriver
``` 

This is due to the chromedriver not having the right permissions, after I add the executable permission it works. However this probably only works with me being sudo user, too. So I'm not sure if we can change the permission of the file inside of python.

uluQulu commented:
@scriptworld-git 
```python
    # downloads the files 
    file = urllib.request.urlopen(url) 
    path = os.path.join(DOWNLOAD_DIR, pf) 
    with open(path, 'wb') as output: 
        output.write(file.read())
```
That block took my attention esp the line opening file with `wb`.
**1**-) `urllib` module works differently with python `2` than `3`
**2**-) That method of writing file to the desired location is a bit dangerous.
`with open(path, 'wb') as output: output.write(file.read());`

#### My suggestions
```python
    from platform import python_version

    # import python version specific `urllib` module component
    if python_version().startswith("3"):
        from urllib.request import urlretrieve
    else:
        from urllib import urlretrieve

    # download the file directly to the assets folder
    urlretrieve(url, path)
```

I will write about other matters, soon, hopefully.


---

Cheers 😁
----

uluQulu commented:
>think that Chromedriver made some huge versioning mistake here :D Shouldn't it be 2.04 ?
It can be thought, though I think they also use string as their versioning value type cos its full version is like `2.42.55893` and cannot be rendered as a floating point number.
But having versions like `2.4.2232` and `2.42.55676` is a bit 🤢

>I learned that this is much better for performance, since python won't load this package into RAM if it doesn't reach the if statement (or function).

There are several reasons to it. E.g. while adding another feature you need to import something, then you do not import it twice in case of already exists. It is needed for fast coding, etc.
Also I don't think those imports will capture more than a few kilo-bytes.

>Downloading the driver takes around 2-5 seconds for me

It is just a few megabytes but the point is reaching ultimate speed in efficiency. And then consider people with 1 Mbit/s [like me] who might wait more than 2-5 seconds 🤫

>If you still want this Feature, you have to be patient tho, since I am really busy with work and my own project sadly... I'm not even using InstaPy this is why I prefer programming for another project :) But I really wanna get this finished in a way that this feature is useable!

Thanks for all the work, as like as you I also do not use InstaPy actively, but I like writing code ☕ and helping people who cannot write.
And sometimes I do some nice stuff using InstaPy. Such as following lots of random photos of friends and adding nicest comments 😁

But you are a fast coder. This improvement at this short time is amazing. I hope you soon finalize this work and it also works at linux and mac.

Let me know if needs any help.

----

Cheers 😁
----

Vivism commented on instapy/webdriver_util.py:
```diff
@@ -0,0 +1,131 @@
+    logger.info("Finished updating webdriver.")
+
+
+def get_latest_chromedriver_version(logger):
```
A GET request to https://chromedriver.storage.googleapis.com/LATEST_RELEASE should suffice.

timgrossmann commented:
Will be replaced by https://github.com/InstaPy/instapy-chromedriver with an automatic cron based update mechanism.
Specific versions and "compatible" will be implemented later on.

To guarantee compatibility with Pi etc. we'll introduce that customly added chromedrivers (in the future workspace folder of the new structure) will have a higher priority than the pip installed one.
=> People can still add chromedrivers with specific versions by hand
